File: prodo100/main.c
#include <stdio.h>

long long boxes[500000][4];
long long arr[500000][2];
int N;

long long min(long long a, long long b);
long long max(long long a, long long b);
int is_intersection(int box1, int box2);
long long calc_area(int box1, int box2);

int cut_boxes();

int main() {
    scanf("%d", &N);

    for (int i = 0; i < N; i ++) {
        scanf("%lld %lld", &arr[i][0], &arr[i][1]);
    }

    int box_index = cut_boxes();

    long long max_area = 0;

    for(int i = 0 ; i < box_index; i++) {
        for(int j = 0; j < box_index; j++) {
            if (i == j) {
                continue;
            }

            if (is_intersection(i,j)) {
                max_area = max(max_area, calc_area(i,j));
            }
        }
    }

    printf("%lld", max_area);


    return 0;
}

int cut_boxes() {
    int box_index = 0;
    long long before_x = 0;

    for(int index = 1 ; index < N-1 ; index++) {
        if (index == 1) {
            before_x = arr[index][0];
            continue;
        }

        // down
        if (arr[index + 1][1] < arr[index][1]) {
            boxes[box_index][0] = before_x;
            boxes[box_index][1] = 0;
            boxes[box_index][2] = arr[index][0];
            boxes[box_index][3] = arr[index][1];
            box_index += 1;
        }

        else if (arr[index+1][1] > arr[index][1]) {
            long long start_x = 0;
            long long end_x = arr[N-1][0];

            if (arr[index-1][1] > arr[index-2][1]) {
                start_x = arr[index-1][0];
            } else {
                int ind = index - 1;

                while (ind > 0) {
                    if (arr[ind-1][1] < arr[index][1] && arr[index][1] < arr[ind][1]) {
                        start_x = arr[ind][0];
                        break;
                    }
                    ind -= 1;
                }
            }

            int ind = index + 1;

            while(ind < N-1) {
                if (arr[ind+1][1] < arr[index][1] && arr[index][1] < arr[ind][1]) {
                    end_x = arr[ind][0];
                    break;
                }

                ind += 1;
            }

            boxes[box_index][0] = start_x;
            boxes[box_index][1] = 0;
            boxes[box_index][2] = end_x;
            boxes[box_index][3] = arr[index][1];
            box_index += 1;
            before_x = arr[index][0];
        }
    }

    return box_index;
}

long long min(long long a, long long b) {
    if (a > b) {
        return b;
    }

    return a;
}

long long max(long long a, long long b) {
    if (a > b) {
        return a;
    }

    return b;
}

int is_intersection(int box1, int box2) {
    if ((boxes[box1][0] < boxes[box2][0] && boxes[box2][0] < boxes[box1][2]) || (boxes[box1][0] < boxes[box2][2] && boxes[box2][2] < boxes[box1][2])) {
        return 1;
    }

    return 0;
}

long long calc_area(int box1, int box2) {
    long long area1 = (boxes[box1][2] - boxes[box1][0]) * (boxes[box1][3] - boxes[box1][1]);
    long long area2 = (boxes[box2][2] - boxes[box2][0]) * (boxes[box2][3] - boxes[box2][1]);

    long long x = min(boxes[box1][2], boxes[box2][2]) - max(boxes[box1][0], boxes[box2][0]);
    long long y = min(boxes[box1][3], boxes[box2][3]) - max(boxes[box1][1], boxes[box2][1]);

    long long intersection_area = x * y;

    return area1 + area2 - intersection_area;
}

File: prodo100/main.py
def user_input():
    n = int(input())

    arr = []

    for _ in range(n):
        point = list(map(int, input().rstrip().split()))
        arr.append(point)

    return n, arr


def cut_box(n, arr):
    boxes = []

    before_x = 0

    for index in range(1, n-1):
        if index == 1:
            before_x = arr[index][0]
            continue

        # down
        if arr[index+1][1] < arr[index][1]:
            boxes.append((before_x, 0, arr[index][0], arr[index][1]))

        # up
        elif arr[index+1][1] > arr[index][1]:
            start_x = 0
            end_x = arr[n-1][0]

            if arr[index-1][1] > arr[index-2][1]:
                start_x = arr[index-1][0]
            else:
                ind = index - 1
                while ind > 0:
                    if arr[ind-1][1] < arr[index][1] < arr[ind][1]:
                        start_x = arr[ind][0]
                        break
                    ind -= 1

            ind = index + 1
            while ind < n-1:
                if arr[ind + 1][1] < arr[index][1] < arr[ind][1]:
                    end_x = arr[ind][0]
                    break
                ind += 1

            boxes.append((start_x, 0, end_x, arr[index][1]))
            before_x = arr[index][0]

    return boxes


def is_intersection(box1, box2):
    return (box1[0] < box2[0] < box1[2]) or (box1[0] < box2[2] < box1[2])


def get_union_area(box1, box2):
    area1 = (box1[2] - box1[0]) * (box1[3] - box1[1])
    area2 = (box2[2] - box2[0]) * (box2[3] - box2[1])

    x = min(box1[2], box2[2]) - max(box1[0], box2[0])
    y = min(box1[3], box2[3]) - max(box1[1], box2[1])

    intersection_area = x * y

    return area1 + area2 - intersection_area


"""
8
0 0
0 20
10 20
10 15
15 15
15 10
20 10
20 0
"""



"""
6
0 0
0 10
10 10
10 20
20 20
20 0
"""

"""
14
0 0
0 30
10 30
10 20
20 20
20 25
30 25
30 30
40 30
40 23
50 23
50 30
60 30
60 0
"""



if __name__ == "__main__":
    n, arr = user_input()

    # print(n, arr)
    boxes = cut_box(n, arr)
    # print(boxes)

    max_area = 0

    for i in range(len(boxes)):
        box1 = boxes[i]
        for j in range(len(boxes)):
            if i == j:
                continue
            box2 = boxes[j]
            if is_intersection(box1, box2):
                max_area = max(max_area, get_union_area(box1, box2))

    print(max_area)



